### PR TITLE
HPCC-22281 Add support DataSource API v1

### PR DESCRIPTION
**Pull request by jpmcmu**

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana please review

**Comment by jpmcmu** [comment]
@rpastrana here is what the user facing API looks like for the keywords specified in HpccOptions
```  
    val df2 = spark.read.format("hpcc")
    .option("host",clusterURL)
    .option("password", password)
    .option("username", username)
    .option("partLimit",100)
    .option("fields","field1, field2")
    .load(fileName)
    
    df2.write.mode(SaveMode.Overwrite)
    .format("hpcc")
    .option("host",clusterURL)
    .option("password", password)
    .option("username", username)
    .option("cluster","mythor")
    .save(fileName + "copy")
```

File: DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccOptions.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    public String projectList = null;
+    public CompressionAlgorithm compression = CompressionAlgorithm.DEFAULT;
+    public int expirySeconds = 120;
+    public int limit = -1;
+
+    public HpccOptions(TreeMap<String,String> parameters) throws Exception
+    {
+        // Extract connection
+        String connectionString = null;
+        if (parameters.containsKey("host")) {
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccOptions.java]: let's apply the format template

File: DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccOptions.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        if (parameters.containsKey("path")) {
+           fileName = (String) parameters.get("path"); 
+        }
+
+        if (parameters.containsKey("cluster")) {
+           clusterName = (String) parameters.get("cluster"); 
+        }
+
+        // Use default value in HpccOptions
+        if (parameters.containsKey("timeout")) {
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccOptions.java]: timeout seems a little ambiguous... fileaccesstimeout? or something better but descriptive?

File: DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccOptions.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            String timeoutStr = (String) parameters.get("timeout");
+            expirySeconds = Integer.parseInt(timeoutStr);
+        }
+
+        if (parameters.containsKey("partlimit")) {
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccOptions.java]: reclimitperhpccfilepart?? maybe that's too wordy.... limitperfilepart?

File: DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccOptions.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            String limitStr = (String) parameters.get("partlimit");
+            limit = Integer.parseInt(limitStr);
+        }
+
+        if (parameters.containsKey("fields")) {
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccOptions.java]: in SQL I'd say "selectlist", in HPCC "projectlist", I'm not sure fields is clear enough in the spark world

File: DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccRelation.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    private SQLContext _sql_context = null;
+    private HpccOptions _options = null;
+
+    public HpccRelation(SQLContext context, HpccOptions options)
+    {
+        _sql_context = context;
+        _options = options;
+    }
+
+    public boolean needsConversion()
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccRelation.java]: is this an override? let's annotate it if so...
Also let's try to add javadoc comments where possible

File: DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccRelation.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+import org.hpccsystems.spark.HpccFile;
+import org.hpccsystems.spark.SparkSchemaTranslator;
+
+public class HpccRelation extends BaseRelation implements PrunedFilteredScan
+{
+    private static Logger log = Logger.getLogger(HpccRelation.class.getName());
+
+    private HpccFile _file = null;
+    private SQLContext _sql_context = null;
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccRelation.java]: I think most of the other classes use camel case instead

File: DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccRelation.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        try {
+            if (_file == null) {
+                _file = new HpccFile(_options.fileName, _options.connectionInfo);
+                _file.setFileAccessExpirySecs(_options.expirySeconds);
+            }
+
+            _file.setRecordLimit(_options.limit);
+
+            if (_options.projectList != null) {
+                projectList = _options.projectList;
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccRelation.java]: does this override the "requiredColumns"?

File: DataAccess/src/main/java/org/hpccsystems/spark/HpccFile.java [hunk 8]
```diff
@@ -90,6 +92,24 @@ public HpccFile(String fileName, Connection espconninfo, String targetColumnList
     super(fileName,espconninfo,targetColumnList,filter,remap_info,maxParts,targetfilecluster);
   }
 
+  /**
+   * Set record limit
+   * @param limit
+   */
+  public void setRecordLimit(int limit)
+  {
+    this.recordLimit = limit;
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/HpccFile.java]: this limit is still a perpart limit right? if so the methodname might mislead...

File: DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccRelation.java (hunk 7 above)
Comment (jpmcmu) [reply DataAccess/src/main/java/org/hpccsystems/spark/datasource/HpccRelation.java]: It does, in my testing it doesn't appear the requiredColumns get used? Or at least it doesn't get used in most circumstances. So I added a projectList option so that we can reliably prune columns.

**Comment by jpmcmu** [comment]
@rpastrana  code review changes have been pushed up. Will squash and rebase from master if it looks good

**Comment by jpmcmu** [comment]
@rpastrana rebased

**Comment by rpastrana** [comment]
@richardkchapman please merge
